src/Hedgehog/Gen.fs

```diff
@@ -242,68 +245,70 @@ module Gen =
                 Random.constant None
             | n ->
                 let r = Random.resize (2 * k + n) r0
-                Random.bind r <| fun x ->
+                Random.bind r (fun x ->
                     if p (Tree.outcome x) then
                         Tree.filter p x |> Some |> Random.constant
                     else
-                        tryN (k + 1) (n - 1)
+                        tryN (k + 1) (n - 1))
 
         Random.sized (tryN 0 << max 1)
 
     /// Generates a value that satisfies a predicate.
     let filter (p : 'a -> bool) (g : Gen<'a>) : Gen<'a> =
         let rec loop () =
-            Random.bind (toRandom g |> tryFilterRandom p) <| function
+            Random.bind (toRandom g |> tryFilterRandom p) (function
                 | None ->
-                    Random.sized <| fun n ->
-                        Random.resize (n + 1) (Random.delay loop)
+                    Random.sized (fun n ->
+                        Random.resize (n + 1) (Random.delay loop))
                 | Some x ->
-                    Random.constant x
+                    Random.constant x)
 
         loop ()
         |> ofRandom
 
     /// Tries to generate a value that satisfies a predicate.
     let tryFilter (p : 'a -> bool) (g : Gen<'a>) : Gen<'a option> =
-        ofRandom << Random.bind (toRandom g |> tryFilterRandom p) <| function
+        let filter = function
             | None ->
-                None |> Tree.singleton |> Random.constant
+                Random.constant (Tree.singleton None)
             | Some x ->
-                Tree.map Some x |> Random.constant
+                Random.constant (Tree.map Some x)
+
+        let flipBind f ma = Random.bind ma f
+
+        toRandom g
+        |> tryFilterRandom p
+        |> flipBind filter
+        |> ofRandom
```

Comment:
Are we sure this one works the same (I don't have a compiler in front of me right now because I'm travelling). The safest would be to just add `( .. )` instead of `<|`.

Reply:
> Are we sure this one works the same

They are the same.  Here is a proof.  All lines are equivalent.
```fsharp
ofRandom << Random.bind gen <| f
(ofRandom << Random.bind gen) <| f
f |> (ofRandom << Random.bind gen)
f |> (Random.bind gen >> ofRandom)
f |> Random.bind gen |> ofRandom
Random.bind gen f |> ofRandom
flip Random.bind f gen |> ofRandom
gen |> flip Random.bind f |> ofRandom
```

Frist,
```fsharp
function
| None -> None |> Tree.singleton
| Some t -> t |> Tree.map Some
```
is also known as `OptionTree.sequence` (according to [how I name nested containers](https://github.com/hedgehogqa/fsharp-hedgehog/pull/260#discussion_r551046128), or `Option.Tree.sequence`).

This can be added in a separate PR.

Second, let's add
```fsharp
[<AutoOpen>]
module internal AutoOpen =
    let flip f b a = f a b
```

This can also be in a separate PR.

Then after completing those PRs and rebasing the branch in this PR, I would implement this as
```fsharp
toRandom g
|> tryFilterRandom p
|> flip Random.bind (OptionTree.sequence >> Random.constant)
|> ofRandom
```

As a reminder the need for `flip` there is being removed in PR #266.

Reply:
I'm definitely in favor of using the `|>` here, since the previous version was very difficult to grok.